### PR TITLE
Find local instance ID using metadata service

### LINKED ISSUE #13556: Wrong OpenStack + Cinder instance IDs
I'm playing with Cinder volume plugin and I noticed that it expects that the machine hostname is the same as the OpenStack instance ID.

In our OpenStack deployment, the hostname is `<instance ID>.long.domain.redhat.com`. Kubernetes OpenStack cloud provider, especially `getComputeIDbyHostname()`, takes this long host name and tries to find an instance with this ID, which obviously fails and Cinder plugin does not mount any volumes.

I don't know if this error is caused just by our local setup or it could be common bug for many OpenStack deployments.

@spothanis, any opinion how it should work?

### PR DESCRIPTION
**Pull request by anguslees**
In particular, this change makes no assumption that the instance ID can
be derived from the local hostname, and so fixes #13556.

Tries in order:
- /var/lib/cloud/data/instance-id (from cloud-init)
- meta_data.json in config drive (requires mounting a disk)
- metadata service (network lookup to http://169.254.169.254/)

This change also fixes Instances.CurrentNodeName to similarly fetch the
local server name from metadata rather than assuming the hostname is
correct.

### CODE REVIEW

**Comment by anguslees** [comment]
This effectively duplicates https://github.com/kubernetes/kubernetes/pull/15537, so probably doesn't need to be reviewed until/if the discussion there decides we prefer this variant.

**Comment by anguslees** [comment]
@googlebot: My employer (Rackspace) has signed the Google corp CLA and I wrote this change on the company's dime.  We have this delightful exchange every time I push a change - it would be great if there was some way to teach you about this.

**Comment by brendandburns** [comment]
ok @anguslees please let me know when/if this is ready to review.

@anguslees regarding the CLA bot, you can create a google group and register it with the CLA bot, ping bburns [at] google.com if you want details about how to do this.

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+			"blkid", "-l",
+			"-t", "LABEL=config-2",
+			"-o", "device",
+		).CombinedOutput()
+		if err != nil {
+			glog.V(2).Infof("Unable to run blkid: %v", err)
+			return Metadata{}, err
+		}
+		dev = strings.TrimSpace(string(out))
+	}
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/openstack/metadata.go]: you need to check `err != nil` here.

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	mntdir, err := ioutil.TempDir("", "configdrive")
+	if err != nil {
+		return Metadata{}, err
+	}
+	defer os.Remove(mntdir)
+
+	glog.V(4).Infof("Attempting to mount configdrive %s on %s", dev, mntdir)
+
+	mounter := mount.New()
+	err = mounter.Mount(dev, mntdir, "iso9660", []string{"ro"})
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/openstack/metadata.go]:
```go
if ...; err != nil {
```

(merge these two lines)

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	if err != nil {
+		err = mounter.Mount(dev, mntdir, "vfat", []string{"ro"})
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/openstack/metadata.go]: please add a comment explaining why (and a glog.V(x).Infof with the error)

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	if err != nil {
+		glog.Errorf("Error mounting configdrive %s: %v", dev, err)
+		return Metadata{}, err
+	}
+	defer mounter.Unmount(mntdir)
+
+	glog.V(4).Infof("Configdrive mounted on %s", mntdir)
+
+	f, err := os.Open(
+		filepath.Join(mntdir, "openstack/2012-08-10/meta_data.json"))
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/openstack/metadata.go]: no need for the line break.

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	defer f.Close()
+
+	return parseMetadata(f)
+}
+
+func getMetadataFromMetadataService() (Metadata, error) {
+	// Try to read instance UUID from metadata service
+	const url = "http://169.254.169.254/openstack/2012-08-10/meta_data.json"
+
+	glog.V(2).Infof("Fetching metadata from %s", url)
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/openstack/metadata.go]: hrm, 2 seems a little spammy, 4 or 6?

File: pkg/cloudprovider/providers/openstack/metadata.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	resp, err := http.Get(url)
+	if err != nil {
+		glog.Errorf("Failed to get metadata from %s: %v", url, err)
+		return Metadata{}, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		glog.Errorf("Metadata service returned %s", resp.Status)
+		return Metadata{}, ErrHttpError
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/openstack/metadata.go]: its probably worth including the statusCode in the error too...

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 7]
```diff
@@ -886,43 +895,30 @@ func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
...
 
-	srvs, err := i.List(".")
-	if err != nil {
-		glog.Errorf("Failed to list servers: %v", err)
-		return "", err
+	// Try to read /var/lib/cloud/data/instance-id, in case
+	// cloud-init has already done the hard work for us.
+	const instanceIDFile = "/var/lib/cloud/data/instance-id"
+	idBytes, err := ioutil.ReadFile(instanceIDFile)
+	if err == nil {
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/openstack/openstack.go]: inline the `if` in the call above.

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 8]
```diff
@@ -886,43 +895,30 @@ func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
...
 	}
+	glog.V(4).Infof("Unable to read instance id from %s: %v",
+		instanceIDFile, err)
 
-	if len(srvs) == 0 {
-		glog.Errorf("Found no servers in the region")
-		return "", errors.New("Found no servers in the region")
+	// Lookup config drive / metadata service our selves.
+	md, err := getMetadata()
+	if err == nil {
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/openstack/openstack.go]: inline the `if` with above.

**Comment by brendandburns** [comment]
Small nits, generally lgtm.

**Comment by alex-mohr** [comment]
Closing as inactive.  Please re-open if you'd like to pursue this!

**Pull request closed without merging**